src/components/VerboseTextFieldArray.tsx

```diff
@@ -0,0 +1,122 @@
+// subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+import {
+  Button,
+  FormHelperText,
+  FormLabel,
+  Grid,
+  TextField,
+} from "@mui/material";
+import { ChangeEventHandler } from "react";
+
+export type VerboseFieldArrayRenderProps = {
+  name: string;
+  label: string;
+  description: string | string[];
+  values: string[];
+
+  onChange?:
+    | ChangeEventHandler<HTMLInputElement | HTMLTextAreaElement>
+    | undefined;
+  onBlur?:
+    | React.FocusEventHandler<HTMLInputElement | HTMLTextAreaElement>
+    | undefined;
+  pushItem(obj: unknown): void;
+  removeItem<T>(index: number): T | undefined;
+};
+
+export default function VerboseTextFieldArray(
+  props: VerboseFieldArrayRenderProps | void
+) {
+  if (!props) {
+    return <Grid container item spacing={1} />;
+  }
+  const {
+    name: fieldName,
+    label,
+    description,
+    values,
+
+    pushItem,
+    removeItem,
+    onBlur,
+    onChange,
+  } = props;
+
+  const descriptions = Array.isArray(description) ? description : [description];
+
+  return (
+    <Grid container item>
+      <Grid container item>
+        <FormLabel>{label}</FormLabel>
+        <Grid item marginLeft={2} paddingRight={2}>
+          {descriptions.map((descriptionParagraph) => (
+            <FormHelperText>{descriptionParagraph}</FormHelperText>
+          ))}
+        </Grid>
+      </Grid>
+
+      <Grid container item spacing={1}>
+        {values.map((redirectUri: string, index: number) => {
+          return (
+            // index={key}, taken from https://formik.org/docs/examples/field-arrays
+            // This seems fine in this context.
+            // eslint-disable-next-line react/no-array-index-key
+            <Grid container item key={index} spacing={1}>
+              <Grid item sx={{ flexGrow: 1 }}>
+                <TextField
+                  name={`${fieldName}.${index}`}
+                  label="Redirect URI"
+                  value={redirectUri}
+                  onBlur={onBlur}
+                  onChange={onChange}
+                  inputProps={{ inputMode: "url" }}
+                  required
+                  fullWidth
+                  size="small"
+                />
+              </Grid>
+              <Grid item>
+                <Button
+                  type="button"
+                  title="Remove Redirect URI of this row"
```

Comment:
This looks very specific to a certain context, which is somewhat misaligned with the component name which suggests it is generic and reusable. Should this be passed down as a prop too?

Reply:
Ah snap! That was not supposed to be there :D